TypeScript conversion of `animation` package

barklund opened the pull request:
## Context

This was a big one 😅 

The `animation` package is quite over-engineered and has a serious lack of DRY-ness with a lot of duplicated code, boilerplate and even logic. This is still only a partial conversion.

At this stage, I would welcome a sanity check, but I actually feel pretty good about the conversion, because I can see all the animations working in Storybook (except for background effects, see todo).

## To-do

### Need-to-have
- [x] Convert to TypeScript
- [x] Move various types and functions around to fix package dependencies
- [x] Invert the dependency between `animation` and `elements`, so `elements` now depend on `animation`
- [x] Update and rewrite tests to TypeScript
- [x] Convert various constants to enums
- [x] Update usage of animation constants in story editor
- [x] Update story editor animation playback
- [x] Fix initial lint issues
- [x] Fix conflicts caused by #12682 (should be trivial, just ignore main and use local only)
- [x] Fix lint errors in Storybook stories (probably just going to exclude them from `tsconfig`)
- [x] Fix test errors (actually only rather few left at this point)
- [x] Fix the last type discrepancy in `createAnimation` (requires making `StoryAnimation` a union type with all the properties specified for each different animation type)

### Bugs in editor
- [x] Delay can sometimes not be changed (e.g. for _Fly in_)
- [x] _Fly in_ does not actually fly in, only fade
- [x] Test all other animations

### Nice-to-have
- [x] Re-enable Storybook stories for animations and update where necessary
- [x] Flatten and simplify effects
- [x] Flatten and simplify parts
- [x] Move simple parts in `/parts/`to `/parts/simple/` and move effects to `/parts/effects/`. They're all only ever referenced by the functions inside `/parts/index.tsx` anyway (which will be split up above).
- [x] Split up contents of `/parts/index.tsx` into dedicated files, and add tests (esp. for `getAnimationEffectFields`)
- [ ] Organize animations defaults in the same style throughout (probably with argument defaults everywhere)
- [ ] Move types around inside the animation package for best possible overview
- [ ] Add Storybook stories for background effects
- [ ] Simplify fields logic
- [ ] Simplify animation parts even further (push all component creation to JIT, not setup)

## User-facing changes

_None!_

## Testing Instructions

_Test everything related to animations including old stories with animation, creating new stories with new animations, previewing animations in the editor, viewing animations in the output story._

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #12207

Commit:
WIP

Commit:
Merge branch 'main' into code/12207-ts-anim

# Conflicts:
#	tsconfig.json

Commit:
WIP: Partial refactor and conversion to TypeScript

Note that the main story editor does not compile at this stage, but the animation package works in isolation and can be seen in action through Storybook: `npm run storybook`.

barklund commented on packages/animation/src/parts/types.ts:
```diff
@@ -0,0 +1,52 @@
+  targetLeafElement?: boolean;
+}
+
+export interface AnimationPart<T extends Keyframes = Keyframes> {
```
This generic seems weird, but is **very important**. It allows complex effects to reach into the keyframes of their constituent animation parts and recompose them into combined keyframes. See how e.g. [`RotateIn` is able to compose the move and spin keyframes](https://github.com/GoogleForCreators/web-stories-wp/pull/12702/files#diff-b261f0c38ed22792a0a8682d753999a7a3052e165f0459a21b48663b4a266327R79-R80). This is only possible because the keyframes property of each part isn't generically types as `Keyframes` but as a specific simple object.

Took me some time to figure out, but it actually just works™️  without too much hassle.

miina replied:
Maybe add that as a code comment, too? 🐶

barklund replied:
Done in 7c0462399a.

Commit:
Forgot to simplify the animation creation in this file

Commit:
Merge branch 'main' into code/12207-ts-anim

# Conflicts:
#	packages/animation/src/components/AMPKeyframes.tsx
#	packages/animation/src/components/context.ts
#	packages/animation/src/components/provider.tsx
#	packages/animation/src/components/useStoryAnimationContext.ts
#	packages/animation/src/effects/backgroundPan/animationProps.js
#	packages/animation/src/effects/fadeIn.ts
#	packages/animation/src/effects/flyIn/animationProps.js
#	packages/animation/src/effects/pulse.ts
#	packages/animation/src/effects/zoom/animationProps.js
#	packages/animation/src/parts/defaultFields.ts
#	packages/animation/src/parts/simpleAnimation.js
#	packages/animation/src/parts/test/index.js
#	packages/animation/src/parts/types.js
#	packages/animation/src/types.js
#	packages/animation/src/types/index.ts
#	packages/animation/src/types/types.js
#	packages/elements/src/utils/index.ts
#	packages/media/src/types/mediaElement.ts

Commit:
Merge branch 'main' into code/12207-ts-anim

swissspidy commented on packages/animation/src/utils/test/sanitizeTimings.ts:
```diff
@@ -0,0 +1,41 @@
+ */
+import sanitizeTimings from '../sanitizeTimings';
+
+describe('sanitizeTimings', () => {
```
Good idea to move these tests to its own file 👍 

Seems like a good opportunity to use `it.each` as well.

```ts
describe('sanitizeTimings', () => {
  it.each([
    [-1, -0],
    [-0, 0],
    [0, 0],
    [1, 1],
    [12, 12],
    [51, 51],
    [500, 500],
    [2000, 2000],
    ])('should sanitize values of `duration` and `delay` animation properties', (value, output) => {
      const { duration, delay } = sanitizeTimings({
        duration: value,
        delay: value,
      });
      expect(duration).toBe(output);
      expect(delay).toBe(output);
  });
});
```

barklund replied:
Good point, done in 6c767c6ece. I haven't actually run these tests, just made sure they didn't throw errors. But I trust they run fine, why wouldn't they? 😂 .

Commit:
Converted tests to TS and added to tsconfig

Commit:
Fixed test for `sanitizeTimings`

Commit:
Fixed a bunch of references to anim enum

A bunch of anim constants have been converted from object to enums and thus had their casing updating from `SCREAMING_SNAKE_CASE` to `PascalCase`.

Commit:
Updated story editor to use new animation state enum

Commit:
Merge branch 'code/12207-ts-anim' of github.com:GoogleForCreators/web-stories-wp into code/12207-ts-anim

# Conflicts:
#	packages/story-editor/src/components/panels/design/animation/animation.js

Commit:
Fixed constant

Commit:
Merge branch 'main' into code/12207-ts-anim

Commit:
Excluded stories from TS

Commit:
Excluded stories from TS

Commit:
Fixed some bugs found in test failures

Commit:
Fixed last ts errors

Thís required creating a union type for all the animations. However, it works a little weird, so had to do a bunch of `as const` stuff in tests, so worry about typing the implementation of this in other packages, but for now this is good.

Commit:
Merge branch 'main' into code/12207-ts-anim

Commit:
Restructured animation parts

Commit:
Merge branch 'main' into code/12207-ts-anim

Commit:
Fixed a missing export

Commit:
Adjusted comment to clarify

Commit:
Fixed bug, that I feel must have been there before too?

Commit:
Fixed bug due to misspelled field name

Yes, `panDirection` would be better, but it is already referenced a ton of places, so it is a lot of unnecessary work to change it.

Commit:
Fixed pulse keyframe generator test

Commit:
Fixed types and test for `getGlobalSpace`

Commit:
Trying to fix errors in d.ts. files

Dunno why eslint complains, but maybe it is the missing header?

Commit:
Fixed test with bad animation ids

Commit:
(Re)fixed test with bad animation ids

miina commented on packages/media/src/types/mediaElement.ts:
```diff
@@ -17,14 +17,16 @@
-  panDir: PropTypes.shape(AnimationInputPropTypes),
-  duration: PropTypes.shape(AnimationInputPropTypes),
-};
+export interface MediaElement extends DimensionableElement {
```
We also have the `MediaElement` in [the `elements` package](https://github.com/GoogleForCreators/web-stories-wp/blob/d01beaa21715ffac5708c9b32cbf8c9322e4d063/packages/elements/src/types/element.ts#L99). Should it be in just one place perhaps?

barklund replied:
Yeah, I was a bit conflicted about that. I didn't know where to put this. I don't want elements in the media package, but I also can't import from the elements package in animations. It's all because of the `getElementOffsets` util function.

This function has been moved around a lot lately - probably too much. Like some troubled teenager in and out of foster homes. I'm starting to think it feels abandoned 😢. Maybe it can find a safe and happy forever-home in the animations package ❤️

miina replied:
Yep, if it's used only for `animations`, maybe it belongs there.

barklund replied:
Moved in 72c630baab.

miina replied:
Hm, do we still need a separate `MediaElement` in the media package (and in the `elements` package)?

miina commented on packages/media/src/getElementOffsets.ts:
```diff
@@ -37,21 +37,22 @@ import {
   FULLBLEED_HEIGHT,
```
This util is only used for media + animation?

barklund replied:
Moved in 72c630baab.

miina commented on packages/animation/src/utils/sanitizeTimings.ts:
```diff
@@ -14,16 +14,23 @@
+  ...rest
+}: AMPEffectTiming): AMPEffectTiming {
+  return {
+    easing: easing || 'linear',
```
Q: Why not assign default instead, e.g. `easing = 'linear', duration = 0, delay = 0`?

barklund replied:
Yeah, that was also one of my todo's, that I forgot again. The animation defaults are defined in three different ways - often all three ways in the same file! Both as argument defaults, as values in a default object, and using logical or in the final object structuring. It's a mess and should be unified. I'll add it.

miina commented on packages/animation/src/utils/getGlobalSpace.ts:
```diff
@@ -15,8 +15,12 @@
  */
+import type { Element } from '../types';
+
+type stringable = string | number | boolean | null | undefined;
```
```suggestion
type Stringable = string | number | boolean | null | undefined;
```

barklund replied:
Fixed in 0e56a96e08.

miina commented on packages/animation/src/types/keyframes.ts:
```diff
@@ -14,11 +14,9 @@
+export type Keyframes = Keyframe[] | PropertyIndexedKeyframes;
 
-export default Context;
+export type Property = string | number | null | undefined;
```
For clarity maybe
```suggestion
export type CSSProperty = string | number | null | undefined;
```

barklund replied:
Fixed in 0e56a96e08 and moved inline (as it's only used in a single component).

miina commented on packages/animation/src/types/element.ts:
```diff
@@ -17,15 +17,18 @@
 
-function useStoryAnimationContext(selector) {
-  return useContextSelector(StoryAnimationContext, selector ?? identity);
+export interface Element extends DimensionableElement {
```
It's a bit confusing currently with `Element` being defined in the elements package as well which also depends on the `units` package. I guess it's like this because you're making the `elements` package depend on the `animation` package and not vice versa?

barklund replied:
Yes, exactly. I could call this one `AnimatableElement` or something like that for clarity? `ElementThatCanBeTheTargetOfAnAnimation`?

miina replied:
If we'd move the orphan util `getElementOffsets` to the animations package, could we then import from the `elements` package instead?

barklund replied:
The `elements` package needs to depend on the `animation` package, as the `elements` package defines the types for a page as well, which contains the `animations` property. We can't reverse this relationship easily. It was reversed earlier, but that was changed to allow this new organization.

miina commented on packages/animation/src/types/effects.ts:
```diff
@@ -0,0 +1,54 @@
+  Float = 'float',
+  Text = 'text',
+  Checkbox = 'checkbox',
+  Range = 'RANGE',
```
Hm, is there a reason for uppercase here?
```suggestion
  Range = 'range',
```

barklund replied:
I didn't even notice that. It was [like that before](https://github.com/GoogleForCreators/web-stories-wp/pull/12702/files#diff-4e09902edc1526f3354d4ff2dd8f68899bfae0a58ecc5e0192e5ac9cfe921647L192-L202), I'll change it.

barklund replied:
Fixed in 99a9703456.

miina commented on packages/animation/src/parts/simple/zoom.ts:
```diff
@@ -17,30 +17,32 @@
-  fill: 'forwards',
-  duration: 1000,
-};
+const defaults: AMPEffectTiming = { fill: 'forwards', duration: 1000 };
```
Nit: `forwards` looks like something that would be nice to move to an enum. Looks like there are only a few available values for `fill` in the code.

barklund replied:
We only have that as a [string union type from the web animations package](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/web-animations-js/index.d.ts#L6), but it will be validated against this union type so we should be safe.

miina commented on packages/animation/src/parts/getAnimationEffectFields.ts:
```diff
@@ -0,0 +1,90 @@
+  let fieldOrder = Object.keys(allFieldsOrder);
+
+  // ZoomAndPand design deviates from the normal input order by putting
+  // a custom field (zoom direction) at the end. This accomodates for that.
```
```suggestion
  // a custom field (zoom direction) at the end. This accommodates for that.
```

barklund replied:
Fixed in 1e4419c7a0.

miina commented on packages/animation/src/parts/emptyAnimationPart.tsx:
```diff
@@ -17,24 +17,29 @@
-};
+function emptyAnimationPart(): AnimationPart {
+  return {
+    id: '-1',
```
Is it `-1` because the type is `string`?

barklund replied:
Note even sure why I did that. I think it was just to be able to see that it worked somewhere. I could be anything, as long as this whole animation part doesn't produce any keyframes.

miina reviewed (commented):
Nice refactor!

Will finish reviewing tomorrow:
<img width="165" alt="Screenshot 2022-11-24 at 16 23 42" src="https://user-images.githubusercontent.com/3294597/203853564-82f1f9f0-4331-4820-b0fe-710e5996ad36.png">

swissspidy committed:
Fix config for text-sets package

swissspidy committed:
Update lock file

swissspidy committed:
Fix types for amp components

swissspidy committed:
Update tsconfig

swissspidy committed:
Try to adjust web-animations-js types

Commit:
Merge branch 'main' into code/12207-ts-anim

Commit:
Removed irrelevant polyfill

Commit:
Moved element offset functions once more

Commit:
Added clarifying comment

Commit:
Fixed some types

Commit:
Fix case

Commit:
Typo

miina reviewed (approved):
LGTM!

Doesn't seem ideal to define a similar element type in different packages but if there's no better alternative currently, then let's go with that.

Commit:
Fixed import after reorg
